File: top_interview_questions/javascript/49_group_anagrams.js
`
@author: Fizz Cao 
@file: 49_group_anagrams.py
@time: 1/10/2021 11:54 pm
`
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    let m = new Map();

    for (let str of strs) {
        let count = new Array(26).fill(0);
        for (let c of str)
            count[c.charCodeAt()-97]++;
        let keyStr = count.join('#')
        m[keyStr] ? m[keyStr].push(str) : m[keyStr] = [str];
    }

    return Object.values(m);

//     let m = new Map();

//     for (let str of strs) {
//         keyStr = str.split("").sort().join("");
//         if (!m.has(keyStr))
//             m.set(keyStr, [])
//         m.set(keyStr, [...m.get(keyStr), str]);
//     }

//     return Array.from(m.values());
};

File: top_interview_questions/python/49_group_anagrams.py
#!/usr/bin/env python
# encoding: utf-8
'''
@author: Fizz Cao 
@file: 49_group_anagrams.py
@time: 1/10/2021 11:54 pm
'''
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ans = defaultdict(list)

        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1
            ans[tuple(count)].append(s)
        return ans.values()

#         ans = defaultdict(list)

#         for s in strs:
#             ans[tuple(sorted(s))].append(s)

#         return ans.values()